src/security/KeyData.cc

```diff
@@ -96,14 +96,34 @@ Security::KeyData::loadX509ChainFromFile()
         return;
     }
 
-    if (X509_check_issued(cert.get(), cert.get()) == X509_V_OK)
-        debugs(83, 5, "Certificate is self-signed, will not be chained");
-    else {
+    if (X509_check_issued(cert.get(), cert.get()) == X509_V_OK) {
+        char *nameStr = X509_NAME_oneline(X509_get_subject_name(cert.get()), nullptr, 0);
+        debugs(83, DBG_IMPORTANT, "Certificate is self-signed, will not be chained: " << nameStr);
+        OPENSSL_free(nameStr);
+    } else {
+        debugs(83, DBG_IMPORTANT, "Using certificate chain in " << certFile);
         // and add to the chain any other certificate exist in the file
+        X509 *lastCert = cert.get();
         while (X509 *ca = PEM_read_bio_X509(bio.get(), nullptr, nullptr, nullptr)) {
-            // XXX: self-signed check should be applied to all certs loaded.
-            // XXX: missing checks that the chained certs are actually part of a chain for validating cert.
-            chain.emplace_front(Security::CertPointer(ca));
+            // get Issuer name of the cert for debug display.
+            char *nameStr = X509_NAME_oneline(X509_get_subject_name(ca), nullptr, 0);
+
+            // self-signed certificates are not valid in a sent chain.
+            if (X509_check_issued(ca, ca) == X509_V_OK) {
+                debugs(83, 5, "CA " << nameStr << " is self-signed, will not be chained: " << nameStr);
+                OPENSSL_free(nameStr);
+                continue;
```

Comment:
Is it possible that we skip or ignore all certificates (here and/or further below)? If yes, should we still check that the last skipped/ignored CA certificate has issued the signing certificate?

Reply:
it is intentional to skip all self-signed CA, irrelevant CA, non-CA entries and chain that are misconfigured/misordered. Each entry loaded is only accepted if it is the next CA in the chain and lastCert/latestCert incremented to always point at the latest chain entry - the one for which the next loop iteration is now seeking a relevant issuer CA.